docker/Makefile

```diff
@@ -0,0 +1,213 @@
+DOCKER_RUN_ARGS += -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer
+DOCKER_RUN_ARGS += --network host
+DOCKER_RUN_COMMAND ?= "stty cols $$(tput cols) rows $$(tput lines) && bash"
+
+stop_contracts :
+	- docker rm -f contracts_container
+
+run_contracts : build_contracts stop_contracts
+	docker run $(DOCKER_RUN_ARGS) -it --name contracts_container pdo_contracts:${CONTRACTS_VERSION} \
+		-c $(DOCKER_RUN_COMMAND)
+
+# -----------------------------------------------------------------
+# -----------------------------------------------------------------
+JUPYTER_SERVICE_HOST ?= $(PDO_HOSTNAME)
+JUPYTER_LEDGER ?= $(PDO_LEDGER_URL)
+JUPYTER_MODE ?= build
+
+DOCKER_JUPYTER_COMMAND = "/project/pdo/tools/start_jupyter.sh"
+DOCKER_JUPYTER_COMMAND += -l $(JUPYTER_LEDGER)
+DOCKER_JUPYTER_COMMAND += --service-host $(JUPYTER_SERVICE_HOST)
+DOCKER_JUPYTER_COMMAND += -m $(JUPYTER_MODE)
+
+stop_jupyter :
+	- docker rm -f jupyter_container
+
+run_jupyter : build_contracts stop_jupyter
+	docker run $(DOCKER_RUN_ARGS) -it --name jupyter_container pdo_contracts:${CONTRACTS_VERSION} \
+		$(DOCKER_JUPYTER_COMMAND)
+
+# -----------------------------------------------------------------
+# We need a repository with the source for the branch we are going
+# to build. In theory this could just be a copy of the local source
+# tree but it might contain other artifacts. So instead we clone
+# the contracts with just a single branch in it. The storage and
+# performance requirements are relatively low.
+# -----------------------------------------------------------------
+contracts :
+	git clone --single-branch --branch $(CONTRACTS_BRANCH) --recurse-submodules $(CONTRACTS_REPO) contracts
+
+clean_contracts :
+	rm -rf contracts
+
+# -----------------------------------------------------------------
+# Testing uses docker-compose to build a set of images that will be configured
+# to run with localhost as the interface. in order to get a clean shutdown we
+# don't use a docker-compose specification for the base, ccf_base and services_base
+# containers. since we are rebuilding the contracts, we force recompilation of these
+# images from the rebuilt contracts. the corresponding images are built using docker
+# build instead of being included in the docker-compose.
+# -----------------------------------------------------------------
+
+# TEST_ARGS provides additional parameters for the compose
+# operations for running the test environment; generally
+# this would set in a make.loc
+TEST_COMPOSE_ARGS += -f test.yaml
+
+build_test : contracts models build_contracts versions
+
+test : clean_config clean_contracts build_test
+	- PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+	 	docker-compose $(TEST_COMPOSE_ARGS) up --abort-on-container-exit
+	PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+	 	docker-compose $(TEST_COMPOSE_ARGS) down
+
+JUPYTER_COMPOSE_ARGS += -f test_jupyter.yaml
+
+test_jupyter : clean_config clean_contracts build_test
```

Comment:
contrary to `test`, `test_jupyter` does not really run an e2e automated test, so i was a bit surprised when it seemed to hang forever until i realized i had to open the browser to manually run some tests and then call shutdown to shut it down.  For manual testing wouldn't `start_jupyter` be enough/better way?  In particular as it also has a target to stop the services ... ;-)

BTW: seems that with nbconvert / nbmake+pytest there are ways to batch-tests notebooks but haven't ever done it myself ...

Reply:
this is correct. there is really now way to test jupyter with an automated test (you can run jupyter notebooks with ipython but our scripts are really designed for interaction).

and you can get a clean shutdown by using the "shutdown server" menu item in the jupyter window (in your browser).

Reply:
> and you can get a clean shutdown by using the "shutdown server" menu item in the jupyter window (in your browser).

I know, i mentioned that. But on a headless test machine it's a bit more painful to do so. Anyway, i managed, just didn't felt it was very intuitive ...

Reply:
i'm running on a headless machine. and it shuts down cleanly. not sure what you mean? the abort on exit in the  compose command should shutdown everything cleanly. as in... i'm running everything remotely, tunneling to my local workstation where i run the browser, i shutdown jupyter through the browser interface and everthing is completely cleaned up.

Reply:
The shutdown i mentioned before i realized that i can do a Ctrl-C and it also cleanly shuts down services.  (i know and mentioned that one can shut it down via browser but but it's a bit cumbersome once you realize that the test_jupyter is not an automated and self-terminating test.  But the Ctrl-C of course also achives that.  And i guess much of my observations will be eventually addressed by the README.md discussed elsewhere and which would proactively address some of the (pervceived) hurdles/misconceptions.

Reply:
documentation updated.